server/service/microsoft_mdm.go

```diff
@@ -1181,17 +1187,238 @@ func (svc *Service) GetMDMWindowsTOSContent(ctx context.Context, redirectUri str
 	return htmlBuf.String(), nil
 }
 
-func (svc *Service) getManagementResponse(ctx context.Context, reqSyncML *fleet.SyncML) (*string, error) {
+// isTrustedRequest checks if the incoming request was sent from MDM enrolled device
+func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML) error {
 	if reqSyncML == nil {
+		return fleet.NewInvalidArgumentError("syncml req message", "message is not present")
+	}
+
+	// TODO - Security checks that need to be implemented
+	// - TLS based auth
+	// - Device auth based on Source/LocURI DeviceID information (this should be present on Enrollment DB)
+
+	return nil
+}
+
+// processIncomingProtocolCommands will process the incoming command
+func (svc *Service) processIncomingProtocolCommands(messageID string, deviceID string, cmd mdm_types.ProtoCmdOperation) (*fleet.SyncMLCmd, error) {
+	// TODO:
+	// - Handle Security Alerts
+	// - Handle Results/Responses
+
+	// We return a status 200 for all the operations
+	return NewSyncMLCmdStatus(messageID, cmd.Cmd.CmdID, cmd.Verb, mdm.CmdStatusOK), nil
+}
+
+// processIncomingMDMCmds process the incoming message from the device
+// It will return the list of operations that need to be sent to the device
+func (svc *Service) processIncomingMDMCmds(deviceID string, reqMsg *fleet.SyncML) ([]*fleet.SyncMLCmd, error) {
+	var responseCmds []*fleet.SyncMLCmd
+
+	// Get the incoming MessageID
+	reqMessageID, err := reqMsg.GetMessageID()
+	if err != nil {
+		return nil, fmt.Errorf("get incoming msg: %w", err)
+	}
+
+	// Acknowledge the message header
+	// msgref is always 0 for the header
+	if err = reqMsg.IsValidHeader(); err == nil {
+		ackMsg := NewSyncMLCmdStatus(reqMessageID, "0", mdm.SyncMLHdrName, mdm.CmdStatusOK)
+		responseCmds = append(responseCmds, ackMsg)
+	}
+
+	// Now we need to check for any operations that need to be processed
+	protoCMDs := reqMsg.GetOrderedCmds()
+
+	// Iterate over the operations and process them
+	for _, protoCMD := range protoCMDs {
+		protoCmd, err := svc.processIncomingProtocolCommands(reqMessageID, deviceID, protoCMD)
+		if err != nil {
+			return nil, fmt.Errorf("process incoming command: %w", err)
+		}
+
+		// Append the operations to the response
+		if (protoCmd != nil) && (protoCmd.IsValid()) {
+			responseCmds = append(responseCmds, protoCmd)
+		}
+	}
+
+	return responseCmds, nil
+}
+
+// getPendingMDMCmds returns the list of pending MDM commands for the device
+func (svc *Service) getPendingMDMCmds(ctx context.Context, deviceID string) ([]*mdm_types.SyncMLCmd, error) {
+	// Getting the list of pending commands
+	// Pending commands will be removed from the DB
+	pendingCmds, err := svc.ds.MDMWindowsGetAndRemovePendingCommands(ctx, deviceID)
+	if err != nil {
+		return nil, fmt.Errorf("getting incoming cmds %w", err)
+	}
```

Comment:
If anything fails past that point, the pending commands have been lost (removed from the DB) with no way to retry. We probably need a more robust approach, unless those pending commands can be safely re-created as needed? Unless I'm missing something, that is not necessarily the case?

Reply:
I agree. Let me know if using a flag to indicate that command was processed should be the way to go here